src/components/MagicCodeInput.js

```diff
@@ -0,0 +1,347 @@
+import _ from 'underscore';
+import styles from '../styles/styles';
+import * as ValidationUtils from '../libs/ValidationUtils';
+import CONST from '../CONST';
+import Text from './Text';
+import TextInput from './TextInput';
+import FormHelpMessage from './FormHelpMessage';
+
+const propTypes = {
+    /** Name attribute for the input */
+    name: PropTypes.string,
+
+    /** Input value */
+    value: PropTypes.string,
+
+    /** Should the input auto focus */
+    autoFocus: PropTypes.bool,
+
+    /** Whether we should wait before focusing the TextInput, useful when using transitions  */
+    shouldDelayFocus: PropTypes.bool,
+
+    /** Error text to display */
+    errorText: PropTypes.string,
+
+    /** Specifies autocomplete hints for the system, so it can provide autofill */
+    autoComplete: PropTypes.oneOf(['sms-otp', 'one-time-code']).isRequired,
+
+    /* Should submit when the input is complete */
+    shouldSubmitOnComplete: PropTypes.bool,
+
+    /** Function to call when the input is changed  */
+    onChangeText: PropTypes.func,
+
+    /** Function to call when the input is submitted or fully complete */
+    onFulfill: PropTypes.func,
+};
+
+const defaultProps = {
+    value: undefined,
+    name: '',
+    autoFocus: true,
+    shouldDelayFocus: false,
+    errorText: '',
+    shouldSubmitOnComplete: true,
+    onChangeText: () => {},
+    onFulfill: () => {},
+};
+
+class MagicCodeInput extends React.PureComponent {
+    constructor(props) {
+        super(props);
+
+        this.inputPlaceholderSlots = Array.from(Array(CONST.MAGIC_CODE_LENGTH).keys());
+        this.inputRefs = {};
+
+        this.state = {
+            input: '',
+            focusedIndex: 0,
+            editIndex: 0,
+            numbers: props.value ? this.decomposeString(props.value) : Array(CONST.MAGIC_CODE_LENGTH).fill(CONST.MAGIC_CODE_EMPTY_CHAR),
+        };
+
+        this.onFocus = this.onFocus.bind(this);
+        this.onChangeText = this.onChangeText.bind(this);
+        this.onKeyPress = this.onKeyPress.bind(this);
+    }
+
+    componentDidMount() {
+        if (!this.props.autoFocus) {
+            return;
+        }
+
+        if (this.props.shouldDelayFocus) {
+            this.focusTimeout = setTimeout(() => this.inputRefs[0].focus(), CONST.ANIMATED_TRANSITION);
+        }
+
+        this.inputRefs[0].focus();
+    }
+
+    componentDidUpdate(prevProps) {
+        if (prevProps.value === this.props.value) {
+            return;
+        }
+
+        this.setState({
+            numbers: this.decomposeString(this.props.value),
+        });
+    }
+
+    componentWillUnmount() {
+        if (!this.focusTimeout) {
+            return;
+        }
+        clearTimeout(this.focusTimeout);
+    }
+
+    /**
+     * Focuses on the input when it is pressed.
+     *
+     * @param {Object} event
+     * @param {Number} index
+     */
+    onFocus(event) {
+        event.preventDefault();
+        this.setState({
+            input: '',
+        });
```

Comment:
Clearing input has caused [this](https://github.com/Expensify/App/issues/18324) regression